usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -4601,6 +4612,245 @@ static CK_RV rsa_ec_generate_keypair(STDLL_TokData_t * tokdata,
+                   __func__, new_pPrivateKeyTemplate[i].type,
+                   new_pPrivateKeyTemplate[i].ulValueLen,
+                   new_ulPrivateKeyAttributeCount);
+    }
+
+    ep11_get_pin_blob(ep11_session,
+                      (ep11_is_session_object
+                       (pPublicKeyTemplate, ulPublicKeyAttributeCount)
+                       || ep11_is_session_object(pPrivateKeyTemplate,
+                                                 ulPrivateKeyAttributeCount)),
+                      &ep11_pin_blob, &ep11_pin_blob_len);
+
+    RETRY_START
+        rc = dll_m_GenerateKeyPair(pMechanism, new_pPublicKeyTemplate,
+                                   new_ulPublicKeyAttributeCount,
+                                   new_pPrivateKeyTemplate,
+                                   new_ulPrivateKeyAttributeCount,
+                                   ep11_pin_blob, ep11_pin_blob_len,
+                                   privkey_blob, &privkey_blob_len, spki,
+                                   &spki_len, ep11_data->target);
+    RETRY_END(rc, tokdata, sess)
+    if (rc != CKR_OK) {
+        rc = ep11_error_to_pkcs11_error(rc, sess);
+        TRACE_ERROR("%s m_GenerateKeyPair rc=0x%lx spki_len=0x%zx "
+                    "privkey_blob_len=0x%zx mech='%s'\n",
+                    __func__, rc, spki_len, privkey_blob_len,
+                    ep11_get_ckm(pMechanism->mechanism));
+        goto error;
+    }
+    TRACE_INFO("%s m_GenerateKeyPair rc=0x%lx spki_len=0x%zx "
+               "privkey_blob_len=0x%zx mech='%s'\n",
+               __func__, rc, spki_len, privkey_blob_len,
+               ep11_get_ckm(pMechanism->mechanism));
+
+    if (spki_len > MAX_BLOBSIZE || privkey_blob_len > MAX_BLOBSIZE) {
+        TRACE_ERROR("%s blobsize error\n", __func__);
+        rc = CKR_KEY_INDIGESTIBLE;
+        goto error;
+    }
+
+    memset(publblob, 0, sizeof(publblob));
+    memset(privblob, 0, sizeof(privblob));
+
+    memcpy(publblob, spki, spki_len);
+    publblobsize = spki_len;
+
+    memcpy(privblob, privkey_blob, privkey_blob_len);
+    privblobsize = privkey_blob_len;
+
+    rc = build_attribute(CKA_IBM_OPAQUE, publblob, publblobsize, &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s build_attribute failed with rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+    rc = template_update_attribute(publ_tmpl, attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s template_update_attribute failed with rc=0x%lx\n",
+                    __func__, rc);
+        goto error;
+    }
+
+    rc = build_attribute(CKA_IBM_OPAQUE, privblob, privblobsize, &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s build_attribute failed with rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+    rc = template_update_attribute(priv_tmpl, attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s template_update_attribute failed with rc=0x%lx\n",
+                    __func__, rc);
+        goto error;
+    }
+
+    /* Decode SPKI */
```

Comment:
Starting from here, the code is specific to the IBM version of the dilithium mechanism and key type.  Should we prepare now for an potentially upcoming PKCS11 dlilithium mechanism and move the extraction of the IBM specific attributes into a separate function, to have it easier in the future? Or should we only care later when this happens?

Reply:
Hmm ... here I would leave the code as is for now. We already renamed the function to ibm_dilithium_generate_keypair(), so we will most likely have a separate routine for an "official" Dilithium standard. But we may definitely put some rework on this routine if IBM would introduce new attributes.

Reply:
agreed.